data-processing/entities/definitions/upload.py

```diff
@@ -1,86 +1,265 @@
+    contexts_by_definition: Dict[EntityId, Context] = {}
+    for entity_summary in processing_summary.entities:
+        entity_id = entity_summary.entity.id_
+        context = entity_summary.context
+        if (entity_id.startswith("definition")) and context is not None:
+            contexts_by_definition[entity_id] = context
+
+    # Upload information for each term.
+    for entity_summary in processing_summary.entities:
+        boxes = [cast(BoundingBox, l) for l in entity_summary.locations]
+        entity = entity_summary.entity
+        context = entity_summary.context
+
+        if not is_textual_term(entity):
+            continue
+
+        term = cast(TermReference, entity)
+
+        # Assemble list of snippets that include this term.
+        contexts_matching_term = contexts_by_term_name.get(term.text, [])
+        snippets = [c.snippet for c in contexts_matching_term]
+        snippet_sentences = [
+            f"{c.tex_path}-{c.sentence_id}" for c in contexts_matching_term
+        ]
+
+        # Create links to the sentences containing definitions for this term.
+        definition_sentences: List[Optional[str]] = []
+        for definition_id in term.definition_ids:
+            if definition_id not in contexts_by_definition:
+                definition_sentences.append(None)
+            definition_context = contexts_by_definition[definition_id]
+            definition_sentences.append(
+                f"{definition_context.tex_path}-{definition_context.sentence_id}"
             )
-            term_infos.append(term_info)
 
-    # Upload definitions before terms, because terms hold references to definitions that can
-    # only be resolved once the definitions have been uploaded.
-    upload_entities(
-        processing_summary.s2_id,
-        processing_summary.arxiv_id,
-        definition_infos,
-        data_version,
-    )
+        term_info = EntityUploadInfo(
+            id_=term.id_,
+            type_="term",
+            bounding_boxes=boxes,
+            data={
+                "name": term.text,
+                "term_type": term.type_ or "unknown",
+                "definitions": term.definitions,
+                "definition_texs": term.definition_texs,
+                "sources": term.sources,
+                "snippets": snippets,
+            },
+            relationships={
+                "sentence": EntityReference(
+                    type_="sentence",
+                    id_=f"{context.tex_path}-{context.sentence_id}"
+                    if context is not None
+                    else None,
+                ),
+                "definition_sentences": [
+                    EntityReference(type_="sentence", id_=id_)
+                    for id_ in definition_sentences
+                ],
+                "snippet_sentences": [
+                    EntityReference(type_="sentence", id_=id_)
+                    for id_ in snippet_sentences
+                ],
```

Comment:
It would be great if you can add some description of each relationship type here. Also, how are they used differently in UI side?

Reply:
Sure thing! I will do this shortly.

`definition_sentences` are IDs of the sentences that contain each of the definitions for a term. They're use to populate the links that take you to the passage where a term is defined.

`snippets` is the list of all other appearances of the term everywhere else in the paper. A `snippet` contains the full text of a sentence where that term appears. `snippet_sentences` is to 'snippets' what 'definition_sentences' is to 'definitions'; specifically, it contains the IDs of each sentence where the term appears elsewhere in the paper, so that we can link from a list of usages of the term to the passage in the paper that contains it.

Reply:
I see. This is clear now. Would it be possible to add a short description of each relationship type, as you described here?

Reply:
Sure thing! Done!